graylog2-server/src/main/java/org/graylog2/grok/GrokPatternServiceImpl.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.grok;
+
+import com.google.common.base.Strings;
+import io.krakens.grok.api.Grok;
+import io.krakens.grok.api.GrokCompiler;
+import io.krakens.grok.api.exception.GrokException;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+abstract class GrokPatternServiceImpl implements GrokPatternService {
```

Comment:
maybe out of scope, but `saveAll` validation also behaves differently in the two implementing classes. Perhaps this could also be moved here?

Reply:
`validateAll` is already moved to the parent. Which part of `saveAll` do you mean?

Reply:
The additional validation of the collections in lines 163-183 in `MongoDbGrokPatternService` and lines 128-135 `InMemoryGrokPatternService` behave differently, but should be consistent in my eyes. Everything else except the actual persisting of the patterns is identical and could be moved to the abstract class.